pkg/crane/copy.go

```diff
@@ -61,6 +56,24 @@ func doCopy(_ *cobra.Command, args []string) {
 		log.Fatalf("getting creds for %q: %v", dstRef, err)
 	}
 
+	// First, try to copy as an index.
+	// If that fails, try to copy as an image.
+	// We have to do things in this order because fallback logic exists in the
+	// registry to convert an index to an  image.
+	// TODO(#388): Figure out which artifact is returned at runtime.
+	idx, err := remote.Index(srcRef, remote.WithAuthFromKeychain(authn.DefaultKeychain))
+	if err == nil {
+		if err := remote.WriteIndex(dstRef, idx, dstAuth, http.DefaultTransport); err == nil {
+			// We succeeded, so return. If we encounter any errors just fall through.
+			// This is less than ideal.
```

Comment:
Yeah :(

This means `crane cp` gets slower in the common case where it's copying an image. Any ideas how to do this better?

Reply:
Yeah but not with our current interfaces:
https://github.com/google/go-containerregistry/issues/388

We need some return type for `remote.Get` or something that is a tagged union so we can collapse the remote image wave function down to either an image or an index, depending on what we get back from the registry.

As it is, this makes "crane cp" slower by a pretty small amount, since it only affects a single copy.